ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -576,7 +576,16 @@ func (r *RayServiceReconciler) shouldPrepareNewRayCluster(ctx context.Context, r
 			}
 		}
 
-		// Case 3: Otherwise, rollout a new cluster.
+		// Case 3: If the KubeRay version has changed and the hashes are not identical, update the RayCluster with the goalClusterHash
+		// and new KubeRay version, but do not restart the RayCluster.
+		activeKubeRayVersion := activeRayCluster.ObjectMeta.Annotations[utils.KubeRayVersion]
+		if activeKubeRayVersion != utils.KUBERAY_VERSION {
+			activeRayCluster.ObjectMeta.Annotations[utils.HashWithoutReplicasAndWorkersToDeleteKey] = goalClusterHash
```

Comment:
I prefer not to update the input argument (`activeRayCluster`) inside this function. How about removing the logic for updating annotations here and directly `return Update` instead?

Reply:
Done in [5b5fe07](https://github.com/ray-project/kuberay/pull/2320/commits/5b5fe078a11a24eabdccd74094020f60ee4510a8), we now just log and call `Update` and the cluster hash and KubeRay version are updated in `constructRayClusterForRayService`.